cli: Detect port errors in rpcconnect and rpcport

tdb3 opened the pull request:
Adds invalid port detection to bitcoin-cli for -rpcconnect and -rpcport.

In addition to detecting malformed/invalid ports (e.g. those outside of the 16-bit port range, not numbers, etc.), bitcoin-cli also now considers usage of port 0 to be invalid.  bitcoin-cli previously considered port 0 to be valid and attempted to use it to reach bitcoind.

Functional tests were added for invalid port detection as well as port prioritization.
Additionally, a warning is provided when a port is specified in both -rpcconnect and -rpcport.

This PR is an alternate approach to PR #27820 (e.g. SplitHostPort is unmodified).

Considered an alternative to 127.0.0.1 being specified in functional tests, but at first glance, this might need an update to test_framework/util.py (e.g.  rpc_url), which might be left to a future PR.

davidgumberg commented on test/functional/interface_bitcoin_cli.py:
```diff
@@ -107,6 +108,23 @@ def run_test(self):
+        assert_raises_process_error(1, "Invalid port provided in -rpcconnect: 127.0.0.1:notaport", self.nodes[0].cli("-rpcconnect=127.0.0.1:notaport").echo)
+        assert_raises_process_error(1, "Invalid port provided in -rpcconnect: 127.0.0.1:-1", self.nodes[0].cli("-rpcconnect=127.0.0.1:-1").echo)
+        assert_raises_process_error(1, "Invalid port provided in -rpcconnect: 127.0.0.1:0", self.nodes[0].cli("-rpcconnect=127.0.0.1:0").echo)
+
```
Nice to have: a check for oversized ports
```diff
+ assert_raises_process_error(1, "Invalid port provided in -rpcconnect: 127.0.0.1:65536", self.nodes[0].cli("-rpcconnect=127.0.0.1:65536").echo)
+
```

Below as well

tdb3 replied:
Thanks.  Added.
Testing upperbound makes sense.

davidgumberg commented on test/functional/interface_bitcoin_cli.py:
```diff
@@ -107,6 +108,23 @@ def run_test(self):
+        assert_equal(BLOCKS, self.nodes[0].cli("-rpcconnect=127.0.0.1:18999", f'-rpcport={node_rpc_port}').getblockcount())
+        assert_equal(BLOCKS, self.nodes[0].cli(f"-rpcconnect=127.0.0.1:{node_rpc_port}").getblockcount())
+        assert_equal(BLOCKS, self.nodes[0].cli(f'-rpcport={node_rpc_port}').getblockcount())
+
```
Also nice to have would be a test when no port is provided that we fall back to `BaseParams().RPCPort()`, it isn't obvious to me how that could be done here though.

tdb3 replied:
Added test for default port in regtest.

Thank you for drawing attention to this!  Looks like explicitly defined `-rpcport` during the creation of the node's bitcoin.conf was causing bitcoin-cli to use the conf-provided rpcport when using rpcconnect without rpcport in command line arguments.  One of the tests (involving `f"-rpcconnect=127.0.0.1:{node_rpc_port}"`) would always pass because of this (i.e. the test wouldn't fail even if a valid but unused port is specified in rpcconnect).  This was verified by creating a temporary test `f"-rpcconnect=127.0.0.1:1234"` that should have failed but didn't.

This was remedied by using `replace_in_config()`.

davidgumberg commented:
crACK https://github.com/bitcoin/bitcoin/pull/29521/commits/46c91a25d93595e3934418b1001fabe906326061

tdb3 commented:
Thanks for the review @davidgumberg!
Added the suggestions (force pushed).
Responses inline.

cbergqvist reviewed (approved):
ACK 6e779d83232025ba83e2bd79bd07d2299410193d

## Testing

Ran `./test/functional/test_runner.py --extended` and got two failures `feature_index_prune.py` and `wallet_importdescriptors.py --descriptors`. Reran on the base of the PR, 1105aa46dd1008c556b8c435f1efcb9be09a1644 and got both failures, so shouldn't be related. Will investigate further.

## Code

I like how you limited the scope of variables introduced in this PR, I wish fewer lines were needed for the added error detection, but I didn't figure out a way to shrink it either.

Some would suggest breaking out to a function, but since it's only used here I agree it's better to inline it.

tdb3 commented on src/bitcoin-cli.cpp:
```diff
@@ -747,8 +747,35 @@ static UniValue CallRPC(BaseRequestHandler* rh, const std::string& strMethod, co
+            }
+
+            // Use the valid port provided
+            port = rpcport_port.value();
```
Good suggestions.  I implemented these changes and added a small comment to increase readability (operator bool of std::optional) is being used.

tdb3 commented:
Thank you for the review @cbergqvist!
Implemented your suggestions and rebased.

rkrux commented on src/bitcoin-cli.cpp:
```diff
@@ -747,8 +747,35 @@ static UniValue CallRPC(BaseRequestHandler* rh, const std::string& strMethod, co
+        if (!SplitHostPort(rpcconnect_str, rpcconnect_port, host)) {
+            throw std::runtime_error(strprintf("Invalid port provided in -rpcconnect: %s", rpcconnect_str));
+        } else {
+            if (rpcconnect_port != 0) {
```
Interesting to find out that why the use of port number 0 is discouraged: https://www.lifewire.com/port-0-in-tcp-and-udp-818145

cbergqvist replied:
Interesting article. Wonder what the behavior is before this PR when specifying 0..

tdb3 replied:
```bash
$ src/bitcoin-cli -rpcport=0 getblockchaininfo
error: timeout on transient error: Could not connect to the server 127.0.0.1:0

Make sure the bitcoind server is running and that you are connecting to the correct RPC port.

$ src/bitcoin-cli -rpcconnect=127.0.0.1:0 getblockchaininfo
error: timeout on transient error: Could not connect to the server 127.0.0.1:0

Make sure the bitcoind server is running and that you are connecting to the correct RPC port.
```

The client tries to reach port 0, which wouldn't be successful.

cbergqvist replied:
Ah yeah, makes sense for the client. 👍

rkrux reviewed (approved):
ACK [cb4f9fc](https://github.com/bitcoin/bitcoin/pull/29521/commits/cb4f9fc5847a7e53fe45d54b1fddf504dee5af82)

Testing Methodolgy:
1. Ran make, which was successful.
2. Ran functional tests, all passed.
3. Tried the following `cli` commands in terminal while the daemon was running:
```
➜  bitcoin git:(20240229_rpcconnectinvalidportdetection) ✗ bcli -rpcconnect=127.0.0.1:notaport help
error: Invalid port provided in -rpcconnect: 127.0.0.1:notaport
➜  bitcoin git:(20240229_rpcconnectinvalidportdetection) ✗ bcli -rpcconnect=127.0.0.1:-1 help
error: Invalid port provided in -rpcconnect: 127.0.0.1:-1
➜  bitcoin git:(20240229_rpcconnectinvalidportdetection) ✗ bcli -rpcconnect=127.0.0.1:65536 help
error: Invalid port provided in -rpcconnect: 127.0.0.1:65536
➜  bitcoin git:(20240229_rpcconnectinvalidportdetection) ✗ bcli -rpcport=notaport help
error: Invalid port provided in -rpcport: notaport
➜  bitcoin git:(20240229_rpcconnectinvalidportdetection) ✗ bcli -rpcport=-1 help
error: Invalid port provided in -rpcport: -1

➜  bitcoin git:(20240229_rpcconnectinvalidportdetection) ✗ bcli -rpcconnect=127.0.0.1:9000 -rpcport=8332 getblockcount
Warning: Port specified in both -rpcconnect and -rpcport. Using -rpcport 8332
448483
```

tdb3 commented:
> ACK [cb4f9fc](https://github.com/bitcoin/bitcoin/pull/29521/commits/cb4f9fc5847a7e53fe45d54b1fddf504dee5af82)

Thank you for the review and testing @rkrux!
Will add a note to the description about usage of port 0 for error detection.

tdb3 commented on src/bitcoin-cli.cpp:
```diff
@@ -747,8 +747,35 @@ static UniValue CallRPC(BaseRequestHandler* rh, const std::string& strMethod, co
     uint16_t port{BaseParams().RPCPort()};
-    SplitHostPort(gArgs.GetArg("-rpcconnect", DEFAULT_RPCCONNECT), port, host);
-    port = static_cast<uint16_t>(gArgs.GetIntArg("-rpcport", port));
+    {
```
yes, it enforces scope.  Not strictly necessary, but will leave it for now.

cbergqvist reviewed (approved):
ACK cb4f9fc5847a7e53fe45d54b1fddf504dee5af82

Verified coverage by changing the new C++ code-block to always throw and ran `test/functional/interface_bitcoin_cli.py`, confirming the exception message, reset to only PR changes and re-ran `test/functional/interface_bitcoin_cli.py` successfully.

Ran `./test/functional/test_runner.py --extended --jobs=16 --timeout-factor=1` and saw the same unrelated tests error out as on the previous version of this PR. Still investigating.

marcofleon reviewed (commented):
Tested ACK. The new error handling code looks good to me upon review. Ran the daemon and gave invalid values for `rpcconnect` and `rpcport` and the proper error messages came up. The behavior of port 0 being invalid is consistent with `src/init.cpp`. Ran `interface_bitcoin_cli.py` too and there were no issues with the added tests.

tdb3 commented:
> Tested ACK. The new error handling code looks good to me upon review. Ran the daemon and gave invalid values for `rpcconnect` and `rpcport` and the proper error messages came up. The behavior of port 0 being invalid is consistent with `src/init.cpp`. Ran `interface_bitcoin_cli.py` too and there were no issues with the added tests.

Thank you for reviewing, @marcofleon!

achow101 commented on test/functional/interface_bitcoin_cli.py:
```diff
@@ -107,6 +108,35 @@ def run_test(self):
+        # prefer rpcport over default
+        assert_equal(BLOCKS, self.nodes[0].cli(f'-rpcport={node_rpc_port}').getblockcount())
+        # use default if nothing else available
+        assert_raises_process_error(1, "Could not connect to the server 127.0.0.1:18443", self.nodes[0].cli("-rpcconnect=127.0.0.1").echo)
```
In caa5242f84a6e1ef90b8a44c1206c7b5b942a00c "cli: Sanitize ports in rpcconnect and rpcport"

This will fail if you just have a regtest node running independent of the test:

```
2024-04-22T22:30:22.751000Z TestFramework (ERROR): Assertion failed
Traceback (most recent call last):
  File "/home/ava/bitcoin/bitcoin/29521/test/functional/test_framework/util.py", line 122, in assert_raises_process_error
    fun(*args, **kwds)
  File "/home/ava/bitcoin/bitcoin/29521/test/functional/test_framework/test_node.py", line 821, in __call__
    return self.cli.send_cli(self.command, *args, **kwargs)
  File "/home/ava/bitcoin/bitcoin/29521/test/functional/test_framework/test_node.py", line 886, in send_cli
    raise subprocess.CalledProcessError(returncode, self.binary, output=cli_stderr)
subprocess.CalledProcessError: Command '/home/ava/bitcoin/bitcoin/29521/src/bitcoin-cli' returned non-zero exit status 1.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/ava/bitcoin/bitcoin/29521/test/functional/test_framework/test_framework.py", line 132, in main
    self.run_test()
  File "/home/ava/bitcoin/bitcoin/29521/test/functional/interface_bitcoin_cli.py", line 136, in run_test
    assert_raises_process_error(1, "Could not connect to the server 127.0.0.1:18443", self.nodes[0].cli("-rpcconnect=127.0.0.1").echo)
  File "/home/ava/bitcoin/bitcoin/29521/test/functional/test_framework/util.py", line 127, in assert_raises_process_error
    raise AssertionError("Expected substring not found:" + e.output)
AssertionError: Expected substring not found:error: Authorization failed: Incorrect rpcuser or rpcpassword
```

tdb3 replied:
Thanks for taking a look and trying it out @achow101. Much appreciated.

Good point. Yes, the test would fail if an independent node is running on 18443.

Although not ideal, do you think this would be reasonable since a warning is provided when bitcoind is running independently? (i.e. `WARNING! There is already a bitcoind process running on this system. Tests may fail unexpectedly due to resource contention!`)

Alternatively, the assert statement could be removed, but that seems a bit lacking, since it seems good to check for default port usage (even if the current check is limited to the default regtest port).

Could also look into refactoring a bit, but that might lead to touching a bit more code than desirable.

Thoughts?

tdb3 replied:
Yes, that sounds reasonable.  Removed the "18443" test and pushed.

tdb3 committed:
cli: Sanitize ports in rpcconnect and rpcport

Adds error handling of invalid ports to rpcconnect and rpcport,
with associated functional tests.

tdb3 committed:
cli: Add warning for duplicate port definition

Adds a warning when both -rpcconnect and -rpcport define
a port to be used.

tdb3 commented on test/functional/interface_bitcoin_cli.py:
```diff
@@ -107,6 +108,33 @@ def run_test(self):
+        assert_raises_process_error(1, "Invalid port provided in -rpcconnect: 127.0.0.1:notaport", self.nodes[0].cli("-rpcconnect=127.0.0.1:notaport").echo)
+        assert_raises_process_error(1, "Invalid port provided in -rpcconnect: 127.0.0.1:-1", self.nodes[0].cli("-rpcconnect=127.0.0.1:-1").echo)
+        assert_raises_process_error(1, "Invalid port provided in -rpcconnect: 127.0.0.1:0", self.nodes[0].cli("-rpcconnect=127.0.0.1:0").echo)
+        assert_raises_process_error(1, "Invalid port provided in -rpcconnect: 127.0.0.1:65536", self.nodes[0].cli("-rpcconnect=127.0.0.1:65536").echo)
```
Thanks @kevkevinpal.  Good example of why review is important.

Your findings nudged me to add some assert test statements to help cover cases involving IPv6 hosts.

We're experiencing the result of the scope of `SplitHostPort()`.  `SplitHostPort()` extracts the host and port.  Although the port is validated, host is not (and the function declaration comment header mentions validating the port, but doesn't mention validating the host).

https://github.com/bitcoin/bitcoin/blob/7fee0ca014b1513e836d2550d1b7512739d5a79a/src/util/strencodings.h#L97-L106

`SplitHostPort()` is treating "127.0.0.1::" as an IPv6 host since "::" is used for IPv6 address shortening.  `SplitHostPort()` detects no port as part of the input string, which would be why `Could not connect to the server 127.0.0.1:::18443` is the error provided for `src/bitcoin-cli -rpcconnect=127.0.0.1:: getblockcount` (the ":18443" is tacked on the end because it's the default port for regtest).

An enhancement would be to perform a validation of the host received from `SplitHostPort()` before bitcoin-cli attempts to contact the host.  I'm thinking we would need to first define what is "valid" and what is "invalid" for use with bitcoin-cli.  The gist of this kind of checking seems to be present in `CNetAddr`.  In general, I would think bitcoin-cli shouldn't be overly restrictive in what it lets the user attempt to use as a host.

https://github.com/bitcoin/bitcoin/blob/7fee0ca014b1513e836d2550d1b7512739d5a79a/src/netaddress.h#L157-L181

For now, it might make the most sense to leave the enhancement of host validation to a new PR since this PR is focused on port error detection.

cbergqvist reviewed (approved):
reACK 867a673bcfd3ca254c306d8cf0c68d4ceb163f97

Inspected `git range-diff cb4f9fc~2..cb4f9fc 867a673~2..867a673`.

Passed `--extended` functional tests (-`feature_dbcrash`, excluded, some others skipped due to configuration). Passed `make check`.

tdb3 commented on test/functional/interface_bitcoin_cli.py:
```diff
@@ -107,6 +109,51 @@ def run_test(self):
+        assert_raises_process_error(1, "Invalid port provided in -rpcconnect: 127.0.0.1:65536", self.nodes[0].cli("-rpcconnect=127.0.0.1:65536").echo)
+
+        self.log.info("Check for ipv6")
+        have_ipv6 = test_ipv6_local()
```
Thanks @cbergqvist.  Good observation.

Looks like rpc_bind.py uses `--ipv4` and `--ipv6` arguments to run the test exclusively with ipv4 or ipv6 (not both), which at first glance doesn't seem to be desirable for interface_bitcoin_cli.py (dependence on network type is a somewhat small part of the test).  I could see there being value in refactoring interface_bitcoin_cli.py to include wider coverage of bitcoin-cli with ipv4 and ipv6, but I'm thinking this sort of refactor would be a good follow-up PR rather than included in the scope of this PR.

In the interim, I did push a minor update to add a log message stating when ipv6 testing is being skipped (and also opportunistically rebased).

tdb3 commented:
Seeing a CI error for arm-linux-gnueabihf-g++ (on a line unchanged in this latest push).  Will look into this.

```
In function ‘UniValue CallRPC(BaseRequestHandler*, const std::string&, const std::vector<std::__cxx11::basic_string<char> >&, const std::optional<std::__cxx11::basic_string<char> >&)’,
    inlined from ‘UniValue ConnectAndCallRPC(BaseRequestHandler*, const std::string&, const std::vector<std::__cxx11::basic_string<char> >&, const std::optional<std::__cxx11::basic_string<char> >&)’ at bitcoin-cli.cpp:908:31:
bitcoin-cli.cpp:764:42: error: ‘result’ may be used uninitialized [-Werror=maybe-uninitialized]
  764 |             if (!rpcport_int.has_value() || rpcport_int.value() == 0) {
      |                 ~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from bitcoin-cli.cpp:23:
./util/strencodings.h: In function ‘UniValue ConnectAndCallRPC(BaseRequestHandler*, const std::string&, const std::vector<std::__cxx11::basic_string<char> >&, const std::optional<std::__cxx11::basic_string<char> >&)’:
./util/strencodings.h:184:7: note: ‘result’ was declared here
  184 |     T result;
      |       ^~~~~~
cc1plus: all warnings being treated as errors
```

cbergqvist commented on src/bitcoin-cli.cpp:
```diff
@@ -743,8 +743,41 @@ static UniValue CallRPC(BaseRequestHandler* rh, const std::string& strMethod, co
+
+        if (std::optional<std::string> rpcport_arg = gArgs.GetArg("-rpcport")) {
+            // -rpcport was specified
+            std::optional<uint16_t> rpcport_int{ToIntegral<uint16_t>(rpcport_arg.value())};
+            if (!rpcport_int.has_value() || rpcport_int.value() == 0) {
```
Yeah, that new version should work. Another possible variant.. but no way to know if the compiler becomes happier:
```suggestion
            const uint16_t rpcport_int{ToIntegral<uint16_t>(rpcport_arg.value()).value_or(0)};
            if (rpcport_int == 0) {
```

tdb3 replied:
Thanks all for the suggestions. Pushed an update.
Since locally I have x86_64 (and not arm), I tried configuring/compiling with `--enable-werror` added to `./configure` (to minimize excess CI churn).  Interestingly enough, this didn't result in the same warning as seen on arm CI.

tdb3 replied:
No luck again with the arm CI.  Will try playing with the specific stage locally more (https://github.com/bitcoin/bitcoin/tree/master/ci).

marcofleon reviewed (commented):
re-ACK e208fb5d3bea4c1fb750cb0028819635ecdeb415. I successfully compiled and built (`make check` passed) on my Arm64 machine but the CI error is related to 32-bit machines if I'm not mistaken so might not be relevant. 

Either way, the port error detection looks good to me. Clear comments and more readable since my last review. I also successfully ran the functional test. I modified the test using valid ports and mixed `-rpcconnect` and `-rpcport` configurations to ensure robustness. The test handled everything correctly.

cbergqvist reviewed (approved):
re ACK 24bc46c83b39149f4845a575a82337eb46d91bdb

Inspected `git range-diff 867a673~2..867a673 24bc46c~2..24bc46c`.

Ran same tests as last time (https://github.com/bitcoin/bitcoin/pull/29521/#pullrequestreview-2046768298) with same results.

Tried to repro a boiled down version of the mysterious compiler warning but was unable to reproduce it. Here is what I ended up with: https://godbolt.org/z/fh5KcbbeW
(Not sure about the legalese around uploading Bitcoin Core source to Godbolt but just pasting `ToIntegral()` shouldn't be that bad).

S3RK commented:
light code review ACK 24bc46c83b39149f4845a575a82337eb46d91bdb

achow101 commented:
ACK 24bc46c83b39149f4845a575a82337eb46d91bdb

pinheadmz commented on src/bitcoin-cli.cpp:
```diff
@@ -743,8 +743,41 @@ static UniValue CallRPC(BaseRequestHandler* rh, const std::string& strMethod, co
+            // If there was a valid port provided in rpcconnect,
+            // rpcconnect_port is non-zero.
+            if (rpcconnect_port != 0) {
+                tfm::format(std::cerr, "Warning: Port specified in both -rpcconnect and -rpcport. Using -rpcport %u\n", port);
```
I'm seeing something a bit unexpected here which is that the IP address passed to `rpcconnect` is still respected even though the port is ignored / overridden by `rpcport`. This came to my attention in the `interface_bitcoin_cli` test while working on another branch where I'd borked ipv6. This line in the test unexpectedly tried to connect to bitcoin on `::1` even though the port passed to `rpcconnect` was junk:

```
assert_equal(BLOCKS, self.nodes[0].cli("-rpcconnect=[::1]:18999", f'-rpcport={node_rpc_port}').getblockcount())
```

pinheadmz replied:
on second thought, this behavior does make sense if a user is crazy enough to pass `IP:port` and then a second `port`
